### ISSUE TITLE
Support for (Time)Distributed layers

### ISSUE THREAD

**Issue by turambar** [eclipse/deeplearning4j#2865]
Keras provides a [TimeDistributed](https://keras.io/layers/wrappers/#timedistributed) layer wrapper that allows the user to apply a layer designed for a rank _K_ input along a particular axis of a rank _K+1_ input. This allows the user to do things like:

* (simple) Apply a Dense layer (rank 1) to every time step of a sequence (rank 2)
* (more complex) Apply the same LSTM (rank 2) to every column of an image (rank 3), then feed the outputs into an LSTM over rows, as in [this example](https://github.com/fchollet/keras/blob/master/examples/mnist_hierarchical_rnn.py) of a [hierarchical RNN](https://arxiv.org/abs/1506.01057).

We'd like to add support for this to DL4J (and not just to facilitate import of such models from Keras). After discussion with @AlexDBlack and @crockpotveggies, here is what I would propose:

* Add `MergeDimensionIntoMinibatchInputPreProcessor` and `UnmergeDimensionFromMinibatchInputPreProcessor` preprocessors that basically generalize, e.g., [RnnToFeedForwardInputPreProcessor](https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/preprocessor/RnnToFeedForwardPreProcessor.java) and [FeedForwardToRnnInputPreProcessor](https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/preprocessor/FeedForwardToRnnPreProcessor.java). These should take at least one constructor argument specifying which dimension to merge into or unmerge out of the minibatch dimension. Maybe should also specify the expected shape of its input.
* Add a `distributeAlongDimension(int)` property to [org.deeplearning4j.nn.conf.layers.Layer](https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/layers/Layer.java) and its Builder (or alternatively, add a `DistributeAlongDimension` Layer "wrapper")
* Make necessary changes to InputPreProcessor inference during model building (MLN [here](https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/MultiLayerConfiguration.java#L417), CG [here](https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/ComputationGraphConfiguration.java#L614)).
* *Might* require adding some new [InputTypes](https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/inputs/InputType.java)

NOTE: a more radical proposal for the future would be to overhaul how we handle InputTypes and InputType inference (for example, we might do away with the idea of hard coding things like "RNN" InputType) and instead just let layers operate on any type of input that is the right shape. One reason to do this is that as we add more exotic data types/layers/preprocessors, we might get a proliferation of InputTypes, making things a little complicated. On the other hand, the current InputType strategy provides a form of "type safety" which is nice.

**Comment by turambar** [comment]
Comments from peanut gallery? @AlexDBlack @crockpotveggies @eraly Also, if we want this done in near future, someone not-me will need to take ownership (I'm snowed under).

**Comment by AlexDBlack** [comment]
Right, current time distributed is via preprocessors. If we are moving to a time-distributed layer, we may not need to add the Merge/Unmerge minibatch preprocessors - and instead, handle that directly in the layer itself.
However, one issue with the time distributed layer approach is the composition of multiple time distributed layers: lstm -> dense -> dense -> lstm for example. In that case, the preprocessor approach (and not the layer wrapper approach) makes more sense, to avoid needless (and possibly costly due to copies) reshaping ops.

As discussed with @turambar, I think revisiting InputType is a good idea... for example, currently there is no distinction between feed forward and "time distributed feed forward" cases of InputType.FeedForward. In other words, we lose that 'is/isn't distributed' metadata, which isn't great.

Better support for variable length inputs (non-fixed image sizes, etc for appropriate architectures) would be good too. And being able to support video data in 5d format would be nice (for CNN + RNN or CNN + pooling on video).

**Comment by turambar** [comment]
I generally concur that this should be handled via preprocessors (vs. a layer wrapper). However, I still think that there needs to be a way to signal whether and how a layer should be time distributed. For example, if an LSTM receives a 4D (minibatch, rows, columns, channels) input, what should it do? Throw an error? Distribute along axis 1 (rows) or 2 (columns)? My preference would be to handle this via a field in the Layer superclass that enables any layer to be (time) distributed along any dimension of an (oversized) input.

Also, another concern: looking at the Keras [hierarchical RNN](https://github.com/fchollet/keras/blob/master/examples/mnist_hierarchical_rnn.py) example, it relies upon the ability to return just the last hidden state (rather than the fully sequence), so that the output of the TimeDistributed LSTM layer is a 128x128 sequence. Can we do this in DL4J currently?

**Comment by turambar** [comment]
Note to self: see discussion on this (now closed) PR as well: https://github.com/deeplearning4j/deeplearning4j/pull/2849

**Comment by turambar** [comment]
Hey look, this was filed as an issue way back last summer: https://github.com/deeplearning4j/deeplearning4j/issues/1231

**Comment by turambar** [comment]
I'm going to take a first stab at this. Will post link to PR as soon as it's ready:
- add a distributeAlongDimension field to Layer base class that controls whether a layer gets time distributed and, if so, along which dimension
- add metadata to InputType that indicate whether it’s a “time distributed” InputType and along which dimension. That should be enough information to decide what to do with it (i.e., you divide dimension 0 by minibatchSize and then put the result in the indicated dimension)
- add preprocessors to take care of the above
- test will be that we should be able to use TimeDistributed machinery to exactly replicate RnnToFeedForward and FeedForwardToRnn preprocessors

@AlexDBlack @huitseeker @crockpotveggies

**Comment by agibsonccc** [comment]
Are we done here yet?

**Comment by turambar** [comment]
@agibsonccc alas, no, and work will proceed slowly (or not at all) for next 48 hours.

**Comment by turambar** [comment]
Returning to this issue this week. Goal is to have it done before next release.

**Comment by jsteggink** [comment]
Hi, I was just wondering, is this still being worked on?

**Comment by crockpotveggies** [comment]
With the advent of Keras 2.0, we will be revisiting all of our Keras integrations this summer. This specific issue will need to be updated since it was originally developed for Keras 1.0. I'd set expectations accordingly. Thanks for inquiring! We're also happy to encourage/guide anyone who'd like to work on it.

**Comment by agibsonccc** [comment]
Implemented long ago.
